restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java

```diff
@@ -891,41 +888,38 @@ void addRowsToMap(Map<String, List<Row>> rowsByDoc, List<Row> rows) {
    * intensive than getFullDocuments.
    */
   public ImmutablePair<JsonNode, ByteBuffer> getFullDocumentsFiltered(
-      Db dbFactory,
       DocumentDB db,
-      String authToken,
       String keyspace,
       String collection,
       List<FilterCondition> filters,
       List<String> fields,
       ByteBuffer initialPagingState,
       int pageSize,
-      int limit,
-      Map<String, String> headers)
+      int limit)
       throws UnauthorizedException {
     ObjectNode docsResult = mapper.createObjectNode();
     LinkedHashSet<String> existsByDoc = new LinkedHashSet<>();
     LinkedHashMap<String, Integer> countsByDoc = new LinkedHashMap<>();
 
-    List<FilterCondition> inCassandraFilters =
+    Optional<FilterCondition> firstSupportedFilter =
         filters.stream()
             .filter(f -> !FilterOp.LIMITED_SUPPORT_FILTERS.contains(f.getFilterOp()))
-            .collect(Collectors.toList());
-    List<FilterCondition> inMemoryFilters =
-        filters.stream()
-            .filter(f -> FilterOp.LIMITED_SUPPORT_FILTERS.contains(f.getFilterOp()))
-            .collect(Collectors.toList());
-
-    Set<String> inCassandraPaths =
-        inCassandraFilters.stream().map(f -> f.getFullFieldPath()).collect(Collectors.toSet());
-    if (inCassandraPaths.size() > 1 || inMemoryFilters.size() > 0) {
-      // If the request involves more than one filter that could be handled by cassandra, or if
-      // there is a combination of supported and unsupported filters, then we have to default to the
-      // I/O intensive in-memory filter because the relevant where clause wouldn't be supported in
-      // CQL (multiple C* filters would indirectly require an `OR` due to the document schema
-      // structure).
-      inCassandraFilters = Collections.emptyList();
-      inMemoryFilters = filters;
+            .findFirst();
+    List<FilterCondition> inCassandraFilters = new ArrayList<>();
+    List<FilterCondition> inMemoryFilters = filters;
+
+    String cassandraPath;
+    if (firstSupportedFilter.isPresent()) {
```

Comment:
So is the optimization that it uses the first supported filter on the server-side then filters the rest in memory? versus before where it just filtered everything in memory.

Reply:
Exactly, yes!